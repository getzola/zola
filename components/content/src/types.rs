use serde::{Deserialize, Serialize};

#[derive(Debug, Copy, Clone, PartialEq, Serialize, Deserialize, Eq)]
#[serde(rename_all = "lowercase")]
pub enum SortBy {
    /// Most recent to oldest
    Date,
    /// Most recent to oldest
    #[serde(rename = "update_date")]
    UpdateDate,
    /// Sort by title lexicographically
    Title,
    /// Sort by titles using the bytes directly
    #[serde(rename = "title_bytes")]
    TitleBytes,
    /// Lower weight comes first
    Weight,
    /// Sort by slug
    Slug,
    /// Sort by permalink
    Permalink,
    /// No sorting
    None,
}
