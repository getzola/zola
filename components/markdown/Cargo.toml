[package]
name = "markdown"
version = "0.1.0"
edition = "2021"
include = ["src/**/*"]

[dependencies]
pest = "2"
pest_derive = "2"

errors = { path = "../errors" }
utils = { path = "../utils" }
config = { path = "../config" }
console = { path = "../console" }
libs = { path = "../libs" }

typst = "0.12.0"
typst-assets = { version = "0.12.0", features = ["fonts"] }
typst-svg = "0.12.0"
time = { version = "0.3.37", features = ["local-offset"] }
flate2 = "1.0.35"
tar = "0.4.43"
ttf-parser = "0.25.1"
urlencoding = "2.1.3"
bincode = "1.3.3"
serde = { version = "1.0.130", features = ["derive"] }
dashmap = { version = "6.1.0", features = ["serde"] }
twox-hash = "2.1.0"
[dev-dependencies]
templates = { path = "../templates" }
insta = "1.12.0"
