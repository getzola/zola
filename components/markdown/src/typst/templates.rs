
pub fn display_math(code: &str) -> String {
    format!(
        r#"
#set page(height: auto, width: auto, margin: 0pt, fill: none)
#set text(14pt)
$ {code} $
"#,
    )
}

pub fn inline_math(code: &str) -> String {
    format!(
        r#"
#set page(height: auto, width: auto, margin: 0pt, fill: none)
#set text(13pt)
#let s = state("t", (:))

#let pin(t) = context {{
    let computed = measure(
        line(length: here().position().y)
    )
    s.update(it => it.insert(t, computed.width) + it)
    }}

#show math.equation: it => {{
    box(it, inset: (top: 0.5em, bottom: 0.5em))
    }}

$pin("l1"){code}$

#context [
    #metadata(s.final().at("l1")) <label>
]
"#,
    )
}

pub fn raw(code: &str) -> String {
    format!(
        r#"
#set page(height: auto, width: auto, margin: 0pt, fill: none)
#set text(16pt)
{code}
"#,
    )
}
