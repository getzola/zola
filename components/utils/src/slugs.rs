use libs::slug;
use serde::{Deserialize, Serialize};

#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Default)]
pub enum SlugifyStrategy {
    /// Classic slugification, the default
    #[default]
    On,
    /// No slugification, only remove unsafe characters for filepaths/urls
    Safe,
    /// Nothing is changed, hope for the best!
    Off,
}

fn strip_chars(s: &str, chars: &str) -> String {
    let mut sanitized_string = s.to_string();
    sanitized_string.retain(|c| !chars.contains(c));
    sanitized_string
}

fn strip_invalid_paths_chars(s: &str) -> String {
    // NTFS forbidden characters : https://gist.github.com/doctaphred/d01d05291546186941e1b7ddc02034d3
    // Also we need to trim whitespaces and `.` from the end of filename
    let trimmed = s.trim_end_matches([' ', '.']);
    strip_chars(trimmed, r#"<>:"/\|?*"#)
}

pub fn slugify_paths(s: &str, strategy: SlugifyStrategy) -> String {
    match strategy {
        SlugifyStrategy::On => slug::slugify(s),
        SlugifyStrategy::Safe => strip_invalid_paths_chars(s),
        SlugifyStrategy::Off => s.to_string(),
    }
}

pub fn slugify_anchors(s: &str, strategy: SlugifyStrategy) -> String {
    match strategy {
        SlugifyStrategy::On => slug::slugify(s),
        SlugifyStrategy::Safe | SlugifyStrategy::Off => {
            s.replace(|c: char| c.is_ascii_whitespace(), "_")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_slugify_paths() {
        let tests = vec![
            // input, (on, safe, off)
            ("input", ("input", "input", "input")),
            ("test ", ("test", "test", "test ")),
            ("tes t", ("tes-t", "tes t", "tes t")),
            // Invalid NTFS
            ("dot. ", ("dot", "dot", "dot. ")),
            ("日本", ("ri-ben", "日本", "日本")),
            ("héhé", ("hehe", "héhé", "héhé")),
            ("test (hey)", ("test-hey", "test (hey)", "test (hey)")),
        ];

        for (input, (on, safe, off)) in tests {
            assert_eq!(on, slugify_paths(input, SlugifyStrategy::On));
            assert_eq!(safe, slugify_paths(input, SlugifyStrategy::Safe));
            assert_eq!(off, slugify_paths(input, SlugifyStrategy::Off));
        }
    }

    #[test]
    fn can_slugify_anchors() {
        let tests = vec![
            // input, (on, safe, off)
            ("input", ("input", "input", "input")),
            ("test ", ("test", "test_", "test_")),
            ("tes t", ("tes-t", "tes_t", "tes_t")),
            // Invalid NTFS
            ("dot. ", ("dot", "dot._", "dot._")),
            ("日本", ("ri-ben", "日本", "日本")),
            ("héhé", ("hehe", "héhé", "héhé")),
            ("test (hey)", ("test-hey", "test_(hey)", "test_(hey)")),
        ];

        for (input, (on, safe, off)) in tests {
            assert_eq!(on, slugify_anchors(input, SlugifyStrategy::On));
            assert_eq!(safe, slugify_anchors(input, SlugifyStrategy::Safe));
            assert_eq!(off, slugify_anchors(input, SlugifyStrategy::Off));
        }
    }
}
