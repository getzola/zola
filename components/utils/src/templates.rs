use std::collections::HashMap;

use libs::tera::{Context, Tera};

use errors::{Result, bail};

const DEFAULT_TPL: &str = include_str!("default_tpl.html");

macro_rules! render_default_tpl {
    ($filename: expr, $url: expr) => {{
        let mut context = Context::new();
        context.insert("filename", $filename);
        context.insert("url", $url);
        Tera::one_off(DEFAULT_TPL, &context, true).map_err(std::convert::Into::into)
    }};
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ShortcodeFileType {
    Markdown,
    Html,
}

#[derive(Debug, Clone)]
pub struct ShortcodeDefinition {
    pub file_type: ShortcodeFileType,
    pub tera_name: String,
}
impl ShortcodeDefinition {
    pub fn new(file_type: ShortcodeFileType, tera_name: &str) -> ShortcodeDefinition {
        let tera_name = tera_name.to_string();

        ShortcodeDefinition { file_type, tera_name }
    }
}

#[derive(Debug, Default, Clone)]
pub struct ShortcodeInvocationCounter {
    amounts: HashMap<String, usize>,
}
impl ShortcodeInvocationCounter {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn get(&mut self, str: &str) -> usize {
        let nth = self.amounts.entry(str.into()).or_insert(0);
        *nth += 1;
        *nth
    }
    pub fn reset(&mut self) {
        self.amounts.clear();
    }
}

/// Fetches all the shortcodes from the Tera instances
pub fn get_shortcodes(tera: &Tera) -> HashMap<String, ShortcodeDefinition> {
    let mut shortcode_definitions = HashMap::new();

    for (identifier, template) in tera.templates.iter() {
        let (file_type, ext_len) = if template.name.ends_with(".md") {
            (ShortcodeFileType::Markdown, "md".len())
        } else {
            (ShortcodeFileType::Html, "html".len())
        };

        if template.name.starts_with("shortcodes/") {
            let head_len = "shortcodes/".len();
            shortcode_definitions.insert(
                identifier[head_len..(identifier.len() - ext_len - 1)].to_string(),
                ShortcodeDefinition::new(file_type, &template.name),
            );
            continue;
        }

        if template.name.starts_with("__zola_builtins/shortcodes/") {
            let head_len = "__zola_builtins/shortcodes/".len();
            let name = &identifier[head_len..(identifier.len() - ext_len - 1)];
            // We don't keep the built-ins one if the user provided one
            if shortcode_definitions.contains_key(name) {
                continue;
            }
            shortcode_definitions
                .insert(name.to_string(), ShortcodeDefinition::new(file_type, &template.name));
        }
    }

    shortcode_definitions
}

/// Renders the given template with the given context, but also ensures that, if the default file
/// is not found, it will look up for the equivalent template for the current theme if there is one.
/// Lastly, if it's a default template (index, section or page), it will just return an empty string
/// to avoid an error if there isn't a template with that name
pub fn render_template(
    name: &str,
    tera: &Tera,
    context: Context,
    theme: &Option<String>,
) -> Result<String> {
    if let Some(template) = check_template_fallbacks(name, tera, theme) {
        return tera.render(template, &context).map_err(std::convert::Into::into);
    }

    // maybe it's a default one?
    match name {
        "index.html" | "section.html" => render_default_tpl!(
            name,
            "https://www.getzola.org/documentation/templates/pages-sections/#section-variables"
        ),
        "page.html" => render_default_tpl!(
            name,
            "https://www.getzola.org/documentation/templates/pages-sections/#page-variables"
        ),
        "single.html" | "list.html" => {
            render_default_tpl!(name, "https://www.getzola.org/documentation/templates/taxonomies/")
        }
        _ => bail!("Tried to render `{}` but the template wasn't found", name),
    }
}

/// Rewrites the path of duplicate templates to include the complete theme path
/// Theme templates  will be injected into site templates, with higher priority for site
/// templates. To keep a copy of the template in case it's being extended from a site template
/// of the same name, we reinsert it with the theme path prepended
pub fn rewrite_theme_paths(tera_theme: &mut Tera, theme: &str) {
    let theme_basepath = format!("{}/templates/", theme);
    let mut new_templates = HashMap::new();
    for (key, template) in &tera_theme.templates {
        let mut tpl = template.clone();
        tpl.name = format!("{}{}", theme_basepath, key);
        new_templates.insert(tpl.name.clone(), tpl);
    }
    // Contrary to tera.extend, hashmap.extend does replace existing keys
    // We can safely extend because there's no conflicting paths anymore
    tera_theme.templates.extend(new_templates);
}

/// Checks for the presence of a given template. If none is found, also looks for a
/// fallback in theme and default templates. Returns the path of the most specific
/// template found, or none if none are present.
pub fn check_template_fallbacks<'a>(
    name: &'a str,
    tera: &'a Tera,
    theme: &Option<String>,
) -> Option<&'a str> {
    // check if it is in the templates
    if tera.templates.contains_key(name) {
        return Some(name);
    }

    // check if it is part of a theme
    if let Some(ref t) = *theme {
        let theme_template_name = format!("{}/templates/{}", t, name);
        if let Some((key, _)) = tera.templates.get_key_value(&theme_template_name) {
            return Some(key);
        }
    }

    // check if it is part of ZOLA_TERA defaults
    let default_name = format!("__zola_builtins/{}", name);
    if let Some((key, _)) = tera.templates.get_key_value(&default_name) {
        return Some(key);
    }

    None
}

#[cfg(test)]
mod tests {
    use crate::templates::{check_template_fallbacks, get_shortcodes};

    use super::rewrite_theme_paths;
    use libs::tera::Tera;

    #[test]
    fn can_rewrite_all_paths_of_theme() {
        let mut tera = Tera::parse("test-templates/*.html").unwrap();
        rewrite_theme_paths(&mut tera, "hyde");
        // special case to make the test work: we also rename the files to
        // match the imports
        for (key, val) in &tera.templates.clone() {
            tera.templates.insert(format!("hyde/templates/{}", key), val.clone());
        }
        // Adding our fake base
        tera.add_raw_template("base.html", "Hello").unwrap();
        tera.build_inheritance_chains().unwrap();

        assert_eq!(
            tera.templates["hyde/templates/index.html"].parent,
            Some("base.html".to_string())
        );
        assert_eq!(
            tera.templates["hyde/templates/child.html"].parent,
            Some("index.html".to_string())
        );
    }

    #[test]
    fn template_fallback_is_successful() {
        let mut tera = Tera::parse("test-templates/*.html").unwrap();
        tera.add_raw_template("hyde/templates/index.html", "Hello").unwrap();
        tera.add_raw_template("hyde/templates/theme-only.html", "Hello").unwrap();

        // Check finding existing template
        assert_eq!(check_template_fallbacks("index.html", &tera, &None), Some("index.html"));

        // Check trying to find non-existent template
        assert_eq!(check_template_fallbacks("not-here.html", &tera, &None), None);

        // Check theme fallback
        assert_eq!(
            check_template_fallbacks("theme-only.html", &tera, &Some("hyde".to_string())),
            Some("hyde/templates/theme-only.html")
        );
    }

    #[test]
    fn can_overwrite_builtin_shortcodes() {
        let mut tera = Tera::parse("test-templates/*.html").unwrap();
        tera.add_raw_template("__zola_builtins/shortcodes/youtube.html", "Builtin").unwrap();
        tera.add_raw_template("shortcodes/youtube.html", "Hello").unwrap();
        let definitions = get_shortcodes(&tera);
        assert_eq!(definitions["youtube"].tera_name, "shortcodes/youtube.html");
    }
}
