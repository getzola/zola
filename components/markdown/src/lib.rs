pub mod cache;
pub mod codeblock;
pub mod context;
pub mod markdown;
pub mod shortcode;
pub mod typst;

use shortcode::{extract_shortcodes, insert_md_shortcodes};

use errors::Result;

use crate::markdown::markdown_to_html;
pub use crate::markdown::Rendered;
pub use context::RenderContext;

pub fn render_content(content: &str, context: &RenderContext) -> Result<markdown::Rendered> {
    // avoid parsing the content if needed
    if !content.contains("{{") && !content.contains("{%") {
        return markdown_to_html(content, context, Vec::new());
    }

    let definitions = context.shortcode_definitions.as_ref();
    // Extract all the defined shortcodes
    let (content, shortcodes) = extract_shortcodes(content, definitions)?;

    // Step 1: we render the MD shortcodes before rendering the markdown so they can get processed
    let (content, html_shortcodes) =
        insert_md_shortcodes(content, shortcodes, &context.tera_context, &context.tera)?;

    // Step 2: we render the markdown and the HTML markdown at the same time
    let html_context = markdown_to_html(&content, context, html_shortcodes)?;

    // TODO: Here issue #1418 could be implemented
    // if do_warn_about_unprocessed_md {
    //     warn_about_unprocessed_md(unprocessed_md);
    // }

    Ok(html_context)
}
