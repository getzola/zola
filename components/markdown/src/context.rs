use std::borrow::Cow;
use std::collections::HashMap;
use std::path::PathBuf;

use config::Config;
use libs::tera::{Context, Tera};
use utils::templates::ShortcodeDefinition;
use utils::types::InsertAnchor;

/// All the information from the zola site that is needed to render HTML from markdown
#[derive(Debug)]
pub struct RenderContext<'a> {
    pub tera: Cow<'a, Tera>,
    pub config: &'a Config,
    pub tera_context: Context,
    pub current_page_path: Option<&'a str>,
    pub parent_absolute: Option<&'a PathBuf>,
    pub current_page_permalink: &'a str,
    pub permalinks: Cow<'a, HashMap<String, String>>,
    pub insert_anchor: InsertAnchor,
    pub lang: &'a str,
    pub shortcode_definitions: Cow<'a, HashMap<String, ShortcodeDefinition>>,
}

impl<'a> RenderContext<'a> {
    pub fn new(
        tera: &'a Tera,
        config: &'a Config,
        lang: &'a str,
        current_page_permalink: &'a str,
        permalinks: &'a HashMap<String, String>,
        insert_anchor: InsertAnchor,
    ) -> RenderContext<'a> {
        let mut tera_context = Context::new();
        tera_context.insert("config", &config.serialize(lang));
        tera_context.insert("lang", lang);

        Self {
            tera: Cow::Borrowed(tera),
            tera_context,
            current_page_path: None,
            current_page_permalink,
            permalinks: Cow::Borrowed(permalinks),
            insert_anchor,
            config,
            lang,
            shortcode_definitions: Cow::Owned(HashMap::new()),
            parent_absolute: None,
        }
    }

    /// Set in another step so we don't add one more arg to new.
    /// And it's only used when rendering pages/section anyway
    pub fn set_shortcode_definitions(&mut self, def: &'a HashMap<String, ShortcodeDefinition>) {
        self.shortcode_definitions = Cow::Borrowed(def);
    }

    /// Same as above
    pub fn set_current_page_path(&mut self, path: &'a str) {
        self.current_page_path = Some(path);
    }

    /// Same as above
    pub fn set_parent_absolute(&mut self, path: &'a PathBuf) {
        self.parent_absolute = Some(path);
    }

    // In use in the markdown filter
    // NOTE: This RenderContext is not i18n-aware, see MarkdownFilter::filter for details
    // If this function is ever used outside of MarkdownFilter, take this into consideration
    pub fn from_config(config: &'a Config) -> RenderContext<'a> {
        Self {
            tera: Cow::Owned(Tera::default()),
            tera_context: Context::new(),
            current_page_path: None,
            current_page_permalink: "",
            permalinks: Cow::Owned(HashMap::new()),
            insert_anchor: InsertAnchor::None,
            config,
            lang: &config.default_language,
            shortcode_definitions: Cow::Owned(HashMap::new()),
            parent_absolute: None,
        }
    }
}
