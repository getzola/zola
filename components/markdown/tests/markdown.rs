use std::collections::HashMap;
use std::sync::Arc;

use libs::tera::Tera;

use config::Config;
use markdown::context::Caches;
use markdown::{render_content, RenderContext};
use templates::ZOLA_TERA;
use utils::slugs::SlugifyStrategy;
use utils::types::InsertAnchor;

mod common;

#[test]
fn can_render_basic_markdown() {
    let cases = vec![
        "Hello world",
        "# Hello world",
        "Hello *world*",
        "Hello\n\tworld",
        "Non rendered emoji :smile:",
        "[a link](image.jpg)",
        "![alt text](image.jpg)",
        "<h1>some html</h1>",
    ];

    let body = common::render(&cases.join("\n")).unwrap().body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_make_zola_internal_links() {
    let body = common::render(
        r#"
[rel link](@/pages/about.md)
[rel link with anchor](@/pages/about.md#cv)
[abs link](https://getzola.org/about/)
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_handle_heading_ids() {
    let mut config = Config::default_for_test();

    let cases = vec![
        // Basic
        "# Hello",
        // Same slug as first
        "# Hello",
        // not a slug because of the slugify strategy chosen
        "# L'Ã©cologie et vous",
        // Chosen slug that already exists without space
        "# Hello{#hello}",
        // Chosen slug that already exists with space
        "# Hello {#hello}",
        "# Hello     {#Something_else}",
        "# Workaround for literal {#â€¦&#125;",
        "# Auto {#*matic*}",
        // and now some empty heading
        "# ",
        "# ",
        // zola internal links
        "# [About](@/pages/about.md)",
        // https://github.com/Keats/gutenberg/issues/297
        "# [Rust](https://rust-lang.org \"Rust homepage\")",
        // and then some markdown in them
        "# `hi`",
        "# *hi*",
        "# **hi**",
        // See https://github.com/getzola/zola/issues/569
        "# text [^1] there\n[^1]: footnote",
        // Chosen slug that already exists with space
        "# Classes {#classes .bold .another}",
    ];
    let body = common::render_with_config(&cases.join("\n"), config.clone()).unwrap().body;
    insta::assert_snapshot!(body);

    // And now test without slugifying everything
    config.slugify.anchors = SlugifyStrategy::Safe;
    let body = common::render_with_config(&cases.join("\n"), config).unwrap().body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_insert_anchors() {
    let cases = vec![
        // Basic
        "# Hello\n# World",
        // https://github.com/Keats/gutenberg/issues/42
        "# Hello!",
        // https://github.com/Keats/gutenberg/issues/53
        "## [Rust](https://rust-lang.org)",
        "# Hello*_()",
    ];
    let body =
        common::render_with_insert_anchor(&cases.join("\n"), InsertAnchor::Left).unwrap().body;
    insta::assert_snapshot!(body);
    let body =
        common::render_with_insert_anchor(&cases.join("\n"), InsertAnchor::Right).unwrap().body;
    insta::assert_snapshot!(body);
    let body =
        common::render_with_insert_anchor(&cases.join("\n"), InsertAnchor::Heading).unwrap().body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_customise_anchor_template() {
    let mut tera = Tera::default();
    tera.extend(&ZOLA_TERA).unwrap();
    tera.add_raw_template("anchor-link.html", " (in {{ lang }})").unwrap();
    let permalinks_ctx = HashMap::new();
    let config = Config::default_for_test();
    let context = RenderContext::new(
        &tera,
        &config,
        &config.default_language,
        "",
        &permalinks_ctx,
        InsertAnchor::Right,
        Arc::new(Caches::default()),
    );
    let body = render_content("# Hello", &context).unwrap().body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_customise_summary_template() {
    let mut tera = Tera::default();
    tera.extend(&ZOLA_TERA).unwrap();
    tera.add_raw_template("summary-cutoff.html", " (in {{ lang }})").unwrap();
    let permalinks_ctx = HashMap::new();
    let config = Config::default_for_test();
    let context = RenderContext::new(
        &tera,
        &config,
        &config.default_language,
        "",
        &permalinks_ctx,
        InsertAnchor::Right,
        Arc::new(Caches::default()),
    );
    let summary = render_content("Hello <!-- more --> World!", &context).unwrap().summary.unwrap();
    insta::assert_snapshot!(summary);
}

#[test]
fn can_use_smart_punctuation() {
    let mut config = Config::default_for_test();
    config.markdown.smart_punctuation = true;
    let body = common::render_with_config(r#"This -- is "it"..."#, config).unwrap().body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_use_definition_list() {
    let mut config = Config::default_for_test();
    config.markdown.definition_list = true;
    let content = r#"
term
: definition
    "#;
    let body = common::render_with_config(content, config).unwrap().body;
    insta::assert_snapshot!(body);
}

#[test]
fn can_use_external_links_class() {
    let mut config = Config::default_for_test();

    // external link class only
    config.markdown.external_links_class = Some("external".to_string());
    let body = common::render_with_config("<https://google.com>", config.clone()).unwrap().body;
    insta::assert_snapshot!("external_link_class", body);

    // internal link (should not add class)
    let body =
        common::render_with_config("[about](@/pages/about.md)", config.clone()).unwrap().body;
    insta::assert_snapshot!("internal_link_no_class", body);

    // reset class, set target blank only
    config.markdown.external_links_class = None;
    config.markdown.external_links_target_blank = true;
    let body = common::render_with_config("<https://google.com>", config.clone()).unwrap().body;
    insta::assert_snapshot!("external_link_target_blank", body);

    // both class and target blank
    config.markdown.external_links_class = Some("external".to_string());
    let body = common::render_with_config("<https://google.com>", config).unwrap().body;
    insta::assert_snapshot!("external_link_class_and_target_blank", body);
}

#[test]
fn can_use_external_links_options() {
    let mut config = Config::default_for_test();

    // no options
    let body = common::render("<https://google.com>").unwrap().body;
    insta::assert_snapshot!("external_link_no_options", body);

    // target blank
    config.markdown.external_links_target_blank = true;
    let body = common::render_with_config("<https://google.com>", config.clone()).unwrap().body;
    insta::assert_snapshot!("external_link_target_blank", body);

    // no follow
    config.markdown.external_links_target_blank = false;
    config.markdown.external_links_no_follow = true;
    let body = common::render_with_config("<https://google.com>", config.clone()).unwrap().body;
    insta::assert_snapshot!("external_link_no_follow", body);

    // no referrer
    config.markdown.external_links_no_follow = false;
    config.markdown.external_links_no_referrer = true;
    let body = common::render_with_config("<https://google.com>", config.clone()).unwrap().body;
    insta::assert_snapshot!("external_link_no_referrer", body);

    // all of them
    config.markdown.external_links_no_follow = true;
    config.markdown.external_links_target_blank = true;
    config.markdown.external_links_no_referrer = true;
    let body = common::render_with_config("<https://google.com>", config).unwrap().body;
    insta::assert_snapshot!("external_link_all_options", body);
}

#[test]
fn can_render_emojis() {
    let mut config = Config::default_for_test();
    config.markdown.render_emoji = true;
    let body = common::render_with_config("Hello, World! :smile:", config).unwrap().body;
    assert_eq!(body, "<p>Hello, World! ðŸ˜„</p>\n");
}

// https://github.com/getzola/zola/issues/747
// https://github.com/getzola/zola/issues/816
#[test]
fn custom_url_schemes_are_untouched() {
    let body = common::render(
        r#"
[foo@bar.tld](xmpp:foo@bar.tld)

[(123) 456-7890](tel:+11234567890)

[blank page](about:blank)
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn all_markdown_features_integration() {
    let body = common::render(
        r#"
<!-- Adapted from https://markdown-it.github.io/ -->

# h1 Heading

## h2 Heading

### h3 Heading

#### h4 Heading

##### h5 Heading

###### h6 Heading

## Horizontal Rules

___

---

***

## Emphasis

**This is bold text**

__This is bold text__

*This is italic text*

_This is italic text_

~~Strikethrough~~


## Blockquotes


> Blockquotes can also be nested...
>> ...by using additional greater-than signs right next to each other...
> > > ...or with spaces between arrows.


## Lists

Unordered

+ Create a list by starting a line with `+`, `-`, or `*`
+ Sub-lists are made by indenting 2 spaces:
  - Marker character change forces new list start:
    * Ac tristique libero volutpat at
    + Facilisis in pretium nisl aliquet
    - Nulla volutpat aliquam velit
+ Very easy!

Ordered

1. Lorem ipsum dolor sit amet
2. Consectetur adipiscing elit
3. Integer molestie lorem at massa


1. You can use sequential numbers...
1. ...or keep all the numbers as `1.`

Start numbering with offset:

57. foo
1. bar


## Code

Inline `code`

Indented code

    // Some comments
    line 1 of code
    line 2 of code
    line 3 of code


Block code "fences"

```
Sample text here...
```

Syntax highlighting

``` js
var foo = function (bar) {
  return bar++;
};

console.log(foo(5));
```

## Shortcodes

## Tables

| Option | Description |
| ------ | ----------- |
| data   | path to data files to supply the data that will be passed into templates. |
| engine | engine to be used for processing templates. Handlebars is the default. |
| ext    | extension to be used for dest files. |

Right aligned columns

| Option | Description |
| ------:| -----------:|
| data   | path to data files to supply the data that will be passed into templates. |
| engine | engine to be used for processing templates. Handlebars is the default. |
| ext    | extension to be used for dest files. |


## Links

[link text](http://duckduckgo.com)

[link with title](http://duckduckgo.com/ "Duck duck go")

## Images

![Minion](https://octodex.github.com/images/minion.png)
![Stormtroopocat](https://octodex.github.com/images/stormtroopocat.jpg "The Stormtroopocat")

Like links, Images also have a footnote style syntax

![Alt text][id]

With a reference later in the document defining the URL location:

[id]: https://octodex.github.com/images/dojocat.jpg  "The Dojocat"

## Smileys

Like :smile:, :cry:

### Footnotes

Footnote 1 link[^first].

Footnote 2 link[^second].

Duplicated footnote reference[^second].

[^first]: Footnote **can have markup**
and multiple paragraphs.

[^second]: Footnote text.
    "#,
    )
    .unwrap()
    .body;
    insta::assert_snapshot!(body);
}

#[test]
fn github_style_footnotes() {
    let mut config = Config::default_for_test();
    config.markdown.bottom_footnotes = true;

    let markdown = r#"This text has a footnote[^1]

[^1]:But it is meaningless.

This text has two[^3] footnotes[^2].

[^2]: not sorted.
[^3]: But they are

[^4]:It's before the reference.

There is footnote definition?[^4]

This text has two[^5] identical footnotes[^5]
[^5]: So one is present.
[^6]: But another in not.

This text has a footnote[^7]

[^7]: But the footnote has another footnote[^8].

[^8]: That's it.

Footnotes can also be referenced with identifiers[^first].

[^first]: Like this: `[^first]`.
"#;

    let body = common::render_with_config(&markdown, config).unwrap().body;
    insta::assert_snapshot!(body);
}

#[test]
fn github_alerts() {
    let mut config = Config::default_for_test();
    config.markdown.github_alerts = true;

    let markdown = r#"
> [!NOTE]  
> alert note

> [!TIP]
> alert tip

> [!IMPORTANT]  
> alert important

> [!WARNING]  
> alert warning

> [!CAUTION]
> alert caution
"#;

    let body = common::render_with_config(&markdown, config).unwrap().body;
    insta::assert_snapshot!(body);
}
