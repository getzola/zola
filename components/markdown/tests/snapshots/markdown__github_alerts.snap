---
source: components/markdown/tests/markdown.rs
expression: body
---
<blockquote class="markdown-alert-note">
<p>alert note</p>
</blockquote>
<blockquote class="markdown-alert-tip">
<p>alert tip</p>
</blockquote>
<blockquote class="markdown-alert-important">
<p>alert important</p>
</blockquote>
<blockquote class="markdown-alert-warning">
<p>alert warning</p>
</blockquote>
<blockquote class="markdown-alert-caution">
<p>alert caution</p>
</blockquote>
